experiment/simulation/js/flipflop.js

```diff
@@ -0,0 +1,780 @@
+            window.componentType = "flipFlop";
+            // deleteElement(this.id);
+            return false;
+        }, false);
+
+        flipFlops[this.id] = this;
+        registerGate(this.id, this);
+    }
+
+    setR(r) {
+        this.r = r;
+    }
+    setS(s) {
+        this.s = s;
+    }
+    setClk(clk) {
+        this.clk = clk;
+    }
+    setQ(q) {
+        this.q = q;
+    }
+    setQbar(qbar) {
+        this.qbar = qbar;
+    }
+    addInputPoints(input) {
+        this.inputPoints.push(input);
+    }
+
+    addOutputPoints(output) {
+        this.outputPoints.push(output);
+    }
+
+    setConnected(val, pos) {
+        if (pos === "Q") {
+            this.qIsConnected = val;
+        }
+        else if (pos === "Q'") {
+            this.qbarIsConnected = val;
+        }
+    }
+
+    generateOutput() {
+
+        // In RS flip flop, when Clock is low, the flip flop is in memory state, hence there is no change in output
+        // When clock is high, if both S and R are low, the flip flop is in memory state
+        // If S is high and R is low, the flip flop is in set state i.e. Q is 1 and Q' is 0
+        // If S is low and R is high, the flip flop is in reset state i.e. Q is 0 and Q' is 1
+        // If both S and R are high, the flip flop gives both Q and Q' as 1 (invalid input generally)
+
+
+        const r = getOutputRS(this.r[0], this.r[1]);
+        const s = getOutputRS(this.s[0], this.s[1]);
+        const clk = getOutputRS(this.clk[0], this.clk[1]);
+
+        if (!clk) {
+            return;
+        }
+        else {
+            if (!s && !r) {
+                return;
+            }
+            else if (!s && r) {
+                this.q = false;
+                this.qbar = true;
+            }
+            else if (s && !r) {
+                this.q = true;
+                this.qbar = false;
+            }
+            else if (s && r) {
+                this.q = true;
+                this.qbar = true;
+            }
+        }
+    }
+}
+
+
+export function addRSFlipFlop(x, y) {
+    const ff = new RSFlipFlop();
+    ff.registerComponent("working-area", x, y);
+}
+
+window.addRSFlipFlop = addRSFlipFlop;
+
+export function getOutputRS(gate, pos) {
+    if (pos === "Q") {
+        return gate.q;
+    }
+    else if (pos === "Q'") {
+        return gate.qbar;
+    }
+    // But if the gate is not an FA, but an input bit, then return the value of the input
+    else {
+        return gate.output
+    }
+}
+
+// done checking
+export function getResultRS(ff) {
+    // check if flipflop type is Gate object
+    if (ff.constructor.name === "Gate") {
+        return;
+    }
+
+    if (getOutputRS(ff.r[0], ff.r[1]) != null && getOutputRS(ff.s[0], ff.s[1]) != null && getOutputRS(ff.clk[0], ff.clk[1]) != null) {
+        ff.generateOutput();
+    }
+    return;
+}
+
+
+// done checking
+export function checkConnectionsRS() {
+    let correctConnection = true;
+    for (let ffID in flipFlops) {

```

Comment:
Please ensure that you need to use the for..in loop and not the for..of loop

Reply:
using for..of in those places instead of for..in throws error saying that the list is not iterable, so in such places we have used for..in only